src/runtime/mod.rs

```diff
@@ -53,46 +61,61 @@ impl Runtime {
     pub fn new(b: &crate::Builder) -> io::Result<Runtime> {
         let rt = tokio::runtime::Builder::new_current_thread()
             .on_thread_park(|| {
-                CURRENT.with(|x| {
-                    let _ = RefCell::borrow_mut(x).uring.submit();
+                CONTEXT.with(|x| {
+                    let _ = x.with_driver(|d| d.uring.submit());
                 });
             })
             .enable_all()
             .build()?;
 
         let local = LocalSet::new();
 
-        let driver = {
-            let _guard = rt.enter();
-            AsyncFd::new(Driver::new(b)?)?
-        };
+        let driver = Driver::new(b)?;
+
+        let driver_fd = driver.as_raw_fd();
 
-        Ok(Runtime { driver, local, rt })
+        CONTEXT.with(|cx| cx.set_driver(driver));
+
+        Ok(Runtime {
+            uring_fd: driver_fd,
+            local,
+            rt,
+        })
     }
 
     /// Runs a future to completion on the current runtime
     pub fn block_on<F>(&self, future: F) -> F::Output
     where
         F: Future,
     {
-        self.driver.get_ref().with(|| {
-            let drive = async {
+        let drive = {
+            let _guard = self.rt.enter();
+            let driver = AsyncFd::new(self.uring_fd).unwrap();
+
+            async move {
                 loop {
                     // Wait for read-readiness
-                    let mut guard = self.driver.readable().await.unwrap();
-                    self.driver.get_ref().tick();
+                    let mut guard = driver.readable().await.unwrap();
+                    CONTEXT.with(|cx| cx.with_driver(|driver| driver.tick()));
                     guard.clear_ready();
                 }
-            };
+            }
+        };
 
-            tokio::pin!(drive);
-            tokio::pin!(future);
+        tokio::pin!(future);
 
-            self.rt
-                .block_on(self.local.run_until(crate::future::poll_fn(|cx| {
-                    assert!(drive.as_mut().poll(cx).is_pending());
-                    future.as_mut().poll(cx)
-                })))
-        })
+        self.local.spawn_local(drive);
```

Comment:
Surprise! This is the biggest change in this PR in my book. I always had a hard time wrapping my head around the earlier block_on workings, what with its assert that gets the drive poll run once before the actual future is polled. And here it is made independent and simply a local task, that I guess will be polled by tokio any time the syncfd becomes ready. Very interesting. Perhaps a significant change. Perhaps very nice.

Reply:
This should hopefully clean things up a fair bit.